File: src/main/java/ch/softappeal/yass/serialize/Writer.java
package ch.softappeal.yass.serialize;

import ch.softappeal.yass.util.Check;
import ch.softappeal.yass.util.Exceptions;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;

public abstract class Writer {

    public abstract void writeByte(byte value) throws Exception;

    public abstract void writeBytes(byte[] buffer, int offset, int length) throws Exception;

    public final void writeBytes(final byte[] buffer) throws Exception {
        writeBytes(buffer, 0, buffer.length);
    }

    /**
     * Writes a short as 2 bytes, high byte first.
     */
    public final void writeShort(final short value) throws Exception {
        writeByte((byte)(value >> 8));
        writeByte((byte)(value >> 0));
    }

    /**
     * Writes an int as 4 bytes, high byte first.
     */
    public final void writeInt(final int value) throws Exception {
        writeByte((byte)(value >> 24));
        writeByte((byte)(value >> 16));
        writeByte((byte)(value >> 8));
        writeByte((byte)(value >> 0));
    }

    /**
     * Writes a long as 8 bytes, high byte first.
     */
    public final void writeLong(final long value) throws Exception {
        writeByte((byte)(value >> 56));
        writeByte((byte)(value >> 48));
        writeByte((byte)(value >> 40));
        writeByte((byte)(value >> 32));
        writeByte((byte)(value >> 24));
        writeByte((byte)(value >> 16));
        writeByte((byte)(value >> 8));
        writeByte((byte)(value >> 0));
    }

    /**
     * Writes a char as 2 bytes, high byte first.
     */
    public final void writeChar(final char value) throws Exception {
        writeByte((byte)(value >> 8));
        writeByte((byte)(value >> 0));
    }

    /**
     * Writes a float as 4 bytes, high byte first.
     */
    public final void writeFloat(final float value) throws Exception {
        writeInt(Float.floatToRawIntBits(value));
    }

    /**
     * Writes a double as 8 bytes, high byte first.
     */
    public final void writeDouble(final double value) throws Exception {
        writeLong(Double.doubleToRawLongBits(value));
    }

    /**
     * Writes an int with <a href="https://developers.google.com/protocol-buffers/docs/encoding#varints">variable length encoding</a>.
     */
    public final void writeVarInt(int value) throws Exception {
        while (true) {
            if ((value & ~0b0111_1111) == 0) {
                writeByte((byte)value);
                return;
            }
            writeByte((byte)((value & 0b0111_1111) | 0b1000_0000));
            value >>>= 7;
        }
    }

    /**
     * Writes a <a href="https://developers.google.com/protocol-buffers/docs/encoding#types">zigzag encoded</a> int with {@link #writeVarInt(int)}.
     */
    public final void writeZigZagInt(final int value) throws Exception {
        writeVarInt((value << 1) ^ (value >> 31));
    }

    /**
     * Writes a long with <a href="https://developers.google.com/protocol-buffers/docs/encoding#varints">variable length encoding</a>.
     */
    public final void writeVarLong(long value) throws Exception {
        while (true) {
            if ((value & ~0b0111_1111) == 0) {
                writeByte((byte)value);
                return;
            }
            writeByte((byte)((value & 0b0111_1111) | 0b1000_0000));
            value >>>= 7;
        }
    }

    /**
     * Writes a <a href="https://developers.google.com/protocol-buffers/docs/encoding#types">zigzag encoded</a> long with {@link #writeVarLong(long)}.
     */
    public final void writeZigZagLong(final long value) throws Exception {
        writeVarLong((value << 1) ^ (value >> 63));
    }

    public final OutputStream stream() {
        return new OutputStream() {
            @Override public void write(final int i) {
                try {
                    writeByte((byte)i);
                } catch (final Exception e) {
                    throw Exceptions.wrap(e);
                }
            }
            @Override public void write(final byte[] b, final int off, final int len) {
                try {
                    writeBytes(b, off, len);
                } catch (final Exception e) {
                    throw Exceptions.wrap(e);
                }
            }
        };
    }

    public static Writer create(final OutputStream out) {
        Check.notNull(out);
        return new Writer() {
            @Override public void writeByte(final byte value) throws IOException {
                out.write(value);
            }
            @Override public void writeBytes(final byte[] buffer, final int offset, final int length) throws IOException {
                out.write(buffer, offset, length);
            }
        };
    }

    public static final class ByteBufferOutputStream extends ByteArrayOutputStream {
        public ByteBufferOutputStream(final int size) {
            super(size);
        }
        public ByteBuffer toByteBuffer() {
            return ByteBuffer.wrap(buf, 0, count);
        }
    }

}

File: kotlin/yass/main/ch/softappeal/yass/serialize/Writer.kt
package ch.softappeal.yass.serialize

import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

abstract class Writer {
    abstract fun writeByte(value: Byte)
    abstract fun writeBytes(buffer: ByteArray, offset: Int, length: Int)
    fun writeBytes(buffer: ByteArray) = writeBytes(buffer, 0, buffer.size)
    fun writeShort(value: Short) {
        writeByte((value.toInt() shr 8).toByte())
        writeByte((value.toInt() shr 0).toByte())
    }

    fun writeInt(value: Int) {
        writeByte((value shr 24).toByte())
        writeByte((value shr 16).toByte())
        writeByte((value shr 8).toByte())
        writeByte((value shr 0).toByte())
    }

    fun writeLong(value: Long) {
        writeByte((value shr 56).toByte())
        writeByte((value shr 48).toByte())
        writeByte((value shr 40).toByte())
        writeByte((value shr 32).toByte())
        writeByte((value shr 24).toByte())
        writeByte((value shr 16).toByte())
        writeByte((value shr 8).toByte())
        writeByte((value shr 0).toByte())
    }

    fun writeChar(value: Char) {
        writeByte((value.toInt() shr 8).toByte())
        writeByte((value.toInt() shr 0).toByte())
    }

    fun writeFloat(value: Float) = writeInt(java.lang.Float.floatToRawIntBits(value))
    fun writeDouble(value: Double) = writeLong(java.lang.Double.doubleToRawLongBits(value))
    fun writeVarInt(v: Int) {
        var value = v
        while (true) {
            if ((value and 0b0111_1111.inv()) == 0) {
                writeByte(value.toByte())
                return
            }
            writeByte(((value and 0b0111_1111) or 0b1000_0000).toByte())
            value = value ushr 7
        }
    }

    fun writeZigZagInt(value: Int) = writeVarInt((value shl 1) xor (value shr 31))
    fun writeVarLong(v: Long) {
        var value = v
        while (true) {
            if ((value and 0b0111_1111.inv()) == 0L) {
                writeByte(value.toByte())
                return
            }
            writeByte(((value and 0b0111_1111) or 0b1000_0000).toByte())
            value = value ushr 7
        }
    }

    fun writeZigZagLong(value: Long) = writeVarLong((value shl 1) xor (value shr 63))
    fun stream() = object : OutputStream() {
        override fun write(i: Int) = writeByte(i.toByte())
        override fun write(b: ByteArray, off: Int, len: Int) = writeBytes(b, off, len)
    }
}

fun writer(out: OutputStream) = object : Writer() {
    override fun writeByte(value: Byte) = out.write(value.toInt())
    override fun writeBytes(buffer: ByteArray, offset: Int, length: Int) = out.write(buffer, offset, length)
}

class ByteBufferOutputStream(size: Int) : ByteArrayOutputStream(size) {
    fun toByteBuffer() = ByteBuffer.wrap(buf, 0, count)
}

fun utf8toBytes(value: String): ByteArray = value.toByteArray(StandardCharsets.UTF_8)
